File: BaekJoon/DP/구간합구하기5_11660.cpp
//
//  구간합구하기5_11660.cpp
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/02/29.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: - 구간합구하기5 11660
// MARK: DP C++ 문제풀이

#if 0
#include <iostream>
#include <vector>
using namespace std;

int main() {
    ios_base :: sync_with_stdio(0); cin.tie(0);
    int N,K; cin>>N>>K;
    vector<vector<int>> G(N+1, vector<int>(N+1, 0));
    for(int i=1; i<=N; i++)
        for(int j=1; j<=N; j++)
            cin>>G[i][j];
    
    for(int i=1; i<=N; i++)
        for(int j=1; j<=N; j++)
            G[i][j] += G[i-1][j] + G[i][j-1] - G[i-1][j-1];
    
    while(K--) {
        int x,y,x2,y2; cin>>x>>y>>x2>>y2;
        int Ans = G[x2][y2] + G[x-1][y-1] - (G[x2][y-1] + G[x-1][y2]);
        printf("%d\n",Ans);
    }
    return 0;
}
#endif

File: BaekJoonPS/BaekjoonAlgorithmPS/DP/구간합구하기5_11660.swift
//
//  구간합구하기5_11660.swift
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/02/29.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: - 구간합구하기 5 11660
// MARK: swift 구간합 TLE 답안

/*
import Foundation

typealias Pair = (Int,Int)

let arr = readLine()!.split(separator: " ").map { Int($0)! }
let N = arr[0], K = arr[1]
var G = [[Int]](repeating: [Int](), count: N+1)

G[0] = [Int](repeating: 0, count: N+1)
for i in 1...N {
    var arr2 = [0]
    let add = readLine()!.split(separator: " ").map { Int($0)! }
    arr2.append(contentsOf: add)
    G[i] = arr2
}

for i in 1...N {
    for j in 1...N {
        G[i][j] = G[i][j] + G[i-1][j] + G[i][j-1] - G[i-1][j-1]
    }
}

for _ in 0..<K {
    let arr3 = readLine()!.split(separator: " ").map { Int($0)! }
    let p: Pair = (arr3[0],arr3[1])
    let p2: Pair = (arr3[2],arr3[3])
    var Ans = G[p2.0][p2.1] - G[p2.0][p.1 - 1] - G[p.0 - 1][p2.1]
    Ans += G[p.0 - 1][p.1 - 1]
    print(Ans)
}
*/
